packages/web3torrent/src/clients/web3torrent-client.ts

```diff
@@ -0,0 +1,124 @@
+        reject(err);
+      } else {
+        resolve(id);
+      }
+    })
+  );
+};
+
+const getStatus = (torrent: ExtendedTorrent, previousStatus: Status): Status => {
+  const {uploadSpeed, downloadSpeed} = torrent;
+  if (previousStatus === Status.Seeding) {
+    return Status.Seeding;
+  }
+  if (torrent.progress && torrent.done) {
+    return Status.Completed;
+  }
+  if (uploadSpeed - downloadSpeed === 0) {
+    return Status.Connecting;
+  }
+  return Status.Downloading;
+};
+
+const getFormattedETA = (torrent: ExtendedTorrent) => {
+  const {done, timeRemaining} = torrent;
+  if (done) {
+    return 'Done';
+  }
+  const remaining = timeRemaining || 0;
+  const days = Math.floor(remaining / (1000 * 60 * 60 * 24));
+  const hours = Math.floor((remaining % (1000 * 60 * 60 * 24)) / (1000 * 60 * 60));
+  const minutes = Math.floor((remaining % (1000 * 60 * 60)) / (1000 * 60));
+  const seconds = Math.floor((remaining % (1000 * 60)) / 1000);
+
+  return timeRemaining === Infinity
+    ? 'ETA Unknown'
+    : `ETA ${(days && days + 'd ') || ''}${(hours && hours + 'h ') || ''}${(minutes &&
+        minutes + 'm ') ||
+        ''}${seconds && seconds + 's'}`;
+};
+
+export const getLiveTorrentData = (previousData: Torrent, infoHash): Torrent => {
+  const baseData: Torrent = {
+    ...previousData,
+    name: previousData.name || 'unknown',
+    status: Status.Connecting
+  };
+
+  if (!infoHash) {
+    // torrent in magnet form
+    return {...baseData, status: Status.Idle};
+  }
+
+  const live = web3torrent.get(infoHash) as ExtendedTorrent;
+  if (!live) {
+    // torrent after being destroyed
+    return {...baseData, destroyed: true, status: Status.Idle};
+  }
+
+  return {
+    // active torrent
+    ...baseData,
+    ...live,
+    ...{
+      name: live.name || baseData.name,
+      length: live.length || baseData.length,
+      downloaded: (live && live.downloaded) || 0,
+      status: getStatus(live, previousData.status),
+      uploadSpeed: live.uploadSpeed,
+      downloadSpeed: live.downloadSpeed,
+      numPeers: live.numPeers,
+      parsedTimeRemaining: getFormattedETA(live),
+      ready: true
+    }
+  };
+};
+
+export const parseMagnetURL: (rawMagnetURL: string) => Torrent = (rawMagnetURL = '') => {
+  const emptyTorrentData = {...EmptyTorrent, magnetURI: rawMagnetURL} as Torrent;
+  if (!rawMagnetURL.trim()) {
+    return emptyTorrentData;
+  }
+  const magnetURI = rawMagnetURL.trim().substring(1);
+  const magnetAsArray = magnetURI.split('&');
+
+  const magnetName = magnetAsArray.find(h => h.includes('dn=') || h.includes('name='));
+  const name = magnetName && magnetName.replace(/(dn=)|(name=)/g, '').replace(/\+/g, ' ');
+
+  const magnetLength = magnetAsArray.find(h => h.includes('xl='));
+  const length: number = Number(magnetLength && magnetLength.replace(/(xl=)/g, ''));
+
+  const magnetCost = magnetAsArray.find(h => h.includes('cost='));
+  const cost: number = Number(magnetCost && magnetCost.replace(/(cost=)/g, ''));
```

Comment:
This could be simplified using [URLSearchParams](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams#Examples):

```ts
const magnetParams = new URLSearchParams(rawMangetUrl.replace(/magnet:/g, ''));
```

As described in the [Magnet URI scheme](https://en.wikipedia.org/wiki/Magnet_URI_scheme#Design), you'd have the following parameters as a maximum possible set:

- `dn` (Display Name): a filename to display to the user, for convenience
- `xl` (eXact Length): size in bytes
- `xt` (eXact Topic): URN containing file hash
- `as` (Acceptable Source): Web link to the file online
- `xs` (eXact Source): P2P link identified by a content-hash
- `kt` (Keyword Topic): a more general search, specifying keywords, rather than a particular file
- `mt` (Manifest Topic): link to the metafile that contains a list of magneto (MAGMA – MAGnet MAnifest)
- `tr` (address TRacker): tracker URL for BitTorrent downloads

Reply:
Of course, we can leave it as a future enhancement, no need to do it right here.